atomicsabi64/atomicsabi64.rst

```diff
@@ -0,0 +1,1161 @@
+
+To ensure that the above Mappings are ABI-compatible we test the compilation of
+Concurrent Programs, where each Atomic Operation is compiled to one of the
+aforementioned Mappings. We test if there is a compiled program that exhibits
+an outcome of execution according to the AArch64 Memory Model contained in §B2
+of the Arm Architecture Reference Manual [ARMARM_] that is not an outcome of
+execution of the source program under the ISO C model. In this section we
+define the process by which we test compatibility.
+
+The Mix Testing Process
+-----------------------
+
+We test for Compiler bugs, a compiler bug is defined as an Outcome of a
+compiled program execution (under the AArch64 model) that is not an Outcome of
+execution of the source Concurrent Program (under the ISO C model). Consider
+the hypothetical example where a source Concurrent Program finishes execution
+in one of three possible outcomes::
+
+  { thread_0:r0=0, thread_1:r0=1 }
+  { thread_0:r0=1, thread_1:r0=0 }
+  { thread_0:r0=1, thread_1:r0=1 }
+
+and one possible compiled program outcome has the following according to the
+AArch64 Memory Model contained in §B2 of the Arm Architecture Reference Manual
+[ARMARM_]::
+
+  { thread_0:X3=0, thread_1:X3=0 } <--- Forbidden by source model, compiler bug!
+  { thread_0:X3=0, thread_1:X3=1 }
+  { thread_0:X3=1, thread_1:X3=0 }
+  { thread_0:X3=1, thread_1:X3=1 }
+
+By comparing ``X3`` and the local variable ``r0`` of the original Concurrent
+Program in this example we see there is one additional outcome of executing the
+compiled program that is not an outcome of executing the source program (under
+the respective models). This suggests the Mappings under question are
+incompatible, and a compiler that implements them exhibits a compiler bug. To
+ensure compatibility we therefore test for the absence of such Outcomes of the
+compiled programs when mixing all combinations of the above Mappings. We define
+the *Mix Testing* process as follows:
+
+#. Given a C/C++ Concurrent Program.
+#. Split it into its representative Atomic Operations.
+#. Compile each Atomic Operation separately using a Compiler Profile that
+   generates Assembly Sequences under a given Mapping.
+#. Combine the Assembly Sequences into *multiple* possible Compiled Programs.
+#. Compute the outcomes of executing the Source Concurrent Program under the
+   ISO C memory model. Get source program outcomes *S*.
+#. Compute the outcomes of each compiled program under the AArch64 memory model
+   [ARMARM_]. Get a *set* of compiled program outcomes *C*.
+#. If any *c* in *C* exhibits a compiler bug with respect to the outcomes *S*
+   then the given mappings are not interoperable.
+
+Using Mix Testing we now define ABI-Compatibility of Atomic Operations.
+
+
+Definition of ABI-Compatibility for Atomic Operations
+-----------------------------------------------------
+
+*A compiler that implements the above set of Mappings is ABI-Compatible with
```

Comment:
As I understand it, the ABI is a binary interface. Tools that implement the interface must implement it correctly for compatibility. With the description here Mix Testing sounds like a compliance test and doesn't necessarily alter the definition of ABI compatibility. It could be argued that if Mix testing is just compliance testing then it doesn't belong in the specification, it would be best described in an adjunct document (for example the design-documents section).

Where I could see it coming in is if an alternative code sequence is to be added then it must be shown to be equivalent using Mix Testing.

Reply:
I think we should explain Mix Testing Technique in the ABI itself. The comment above ([here](https://github.com/ARM-software/abi-aa/pull/256#discussion_r1555923959) demonstrates why quite well. By this I mean to help answer your question:

> "(could it be multiple compiler profiles?) [...] Stitch together the assembly outputs from above into multiple (all?) possible (legal) interleavings of execution that when run produce C."

The Mix testing Technique compiles all source fragments using all input profiles, and then generate all the tests with the compiled sequences. Since a given Atomic operation has multiple mappings, there are multiple possible assembly sequences we could use. Only once this stitching is done do we check the executions of _all_ tests (1 input test : multiple output tests). We should document exactly what it is we are doing (mix testing) in the ABI so these same questions don't arise in the reader later on. Compatibility is deliberately defined in terms of Mix Testing because the latter is the technique by which we ensure the former (as far as testing can ensure).

Atomics are complicated as Wilco has suggested in other comments. The technique is equally tricky. I think stating exactly _how_ ABI compatibility is achieved in the same document is important to ensure there is no confusion.

Note also this is independent of discussions of any tool that might implement the mappings (compiler) or mix testing (a testing tool) which I suppose is where a design document comes in.

Reply:
perhaps a middle ground is the appendix, and I'm happy to provide even more documentation, but Mix Testing certainly should remain in this doc.